File: examples/destination_connector/python/table_metadata_helper.py
import sys
sys.path.append('sdk_pb2')

from sdk_pb2 import common_pb2

# Constants for system columns
FIVETRAN_START = "_fivetran_start"
FIVETRAN_END = "_fivetran_end"
FIVETRAN_ACTIVE = "_fivetran_active"

INFO = "INFO"
WARNING = "WARNING"


class TableMetadataHelper:
    """Helper class for table metadata operations"""

    @staticmethod
    def create_table_copy(table_obj, new_name):
        """Creates a copy of a table."""
        table_copy = table_obj.__class__.FromString(table_obj.SerializeToString())
        if hasattr(table_copy, "name"):
            table_copy.name = new_name
        return table_copy

    @staticmethod
    def remove_column_from_table(table_obj, column_name):
        """Removes a column from a table."""
        if not column_name or not hasattr(table_obj, 'columns'):
            return
        # Create a new list of columns excluding the specified column
        columns_to_keep = [col for col in table_obj.columns if col.name != column_name]
        # Clear and repopulate
        del table_obj.columns[:]
        table_obj.columns.extend(columns_to_keep)

    @staticmethod
    def add_history_mode_columns(table_obj):
        """Adds history mode columns to a table."""
        if not hasattr(table_obj, 'columns'):
            return
        start_col = table_obj.columns.add()
        start_col.name = FIVETRAN_START
        start_col.type = common_pb2.DataType.UTC_DATETIME

        end_col = table_obj.columns.add()
        end_col.name = FIVETRAN_END
        end_col.type = common_pb2.DataType.UTC_DATETIME

        active_col = table_obj.columns.add()
        active_col.name = FIVETRAN_ACTIVE
        active_col.type = common_pb2.DataType.BOOLEAN

    @staticmethod
    def add_history_mode_columns_to_db(db_helper, schema, table):
        """Adds history mode columns to a table in the database."""
        columns_to_add = [
            common_pb2.Column(name=FIVETRAN_START, type=common_pb2.DataType.UTC_DATETIME),
            common_pb2.Column(name=FIVETRAN_END, type=common_pb2.DataType.UTC_DATETIME),
            common_pb2.Column(name=FIVETRAN_ACTIVE, type=common_pb2.DataType.BOOLEAN)
        ]

        for column in columns_to_add:
            try:
                db_helper.add_column(schema, table, column)
            except Exception as e:
                log_message(WARNING, f"Failed to add column {column.name} to {schema}.{table}: {str(e)}")

    @staticmethod
    def remove_history_mode_columns_from_db(db_helper, schema, table):
        """Removes history mode columns from a table in the database."""
        columns_to_drop = [FIVETRAN_START, FIVETRAN_END, FIVETRAN_ACTIVE]

        for column in columns_to_drop:
            try:
                db_helper.drop_column(schema, table, column)
            except Exception as e:
                log_message(WARNING, f"Failed to drop column {column} from {schema}.{table}: {str(e)}")

    @staticmethod
    def add_soft_delete_column_to_db(db_helper, schema, table, column_name):
        """Adds a soft delete column to a table in the database."""
        if column_name:
            soft_del_col = common_pb2.Column(name=column_name, type=common_pb2.DataType.BOOLEAN)
            db_helper.add_column(schema, table, soft_del_col)


def log_message(level, message):
    import json
    log_entry = {"level": level, "message": json.dumps(message), "message-origin": "sdk_destination"}
    print(json.dumps(log_entry))

File: examples/destination_connector/java/src/main/java/destination/TableMetadataHelper.java
package destination;

import com.google.gson.Gson;
import fivetran_sdk.v2.Column;
import fivetran_sdk.v2.DataType;
import fivetran_sdk.v2.Table;

import java.sql.SQLException;
import java.util.*;

/**
 * Helper class for table metadata operations.
 */
public class TableMetadataHelper {
    private static final String INFO = "INFO";
    private static final String WARNING = "WARNING";

    // Constants for system columns
    public static final String FIVETRAN_START = "_fivetran_start";
    public static final String FIVETRAN_END = "_fivetran_end";
    public static final String FIVETRAN_ACTIVE = "_fivetran_active";

    /**
     * Creates a copy of a table.
     */
    public static Table createTableCopy(Table tableObj, String newName) {
        return Table.newBuilder(tableObj)
            .setName(newName)
            .build();
    }

    /**
     * Removes a column from a table.
     */
    public static Table removeColumnFromTable(Table tableObj, String columnName) {
        if (columnName == null || columnName.isEmpty()) {
            return tableObj;
        }

        Table.Builder builder = Table.newBuilder(tableObj);
        builder.clearColumns();

        for (Column col : tableObj.getColumnsList()) {
            if (!col.getName().equals(columnName)) {
                builder.addColumns(col);
            }
        }

        return builder.build();
    }

    /**
     * Adds history mode columns to a table.
     */
    public static Table addHistoryModeColumns(Table tableObj) {
        Table.Builder builder = Table.newBuilder(tableObj);

        Column startCol = Column.newBuilder()
            .setName(FIVETRAN_START)
            .setType(DataType.UTC_DATETIME)
            .build();

        Column endCol = Column.newBuilder()
            .setName(FIVETRAN_END)
            .setType(DataType.UTC_DATETIME)
            .build();

        Column activeCol = Column.newBuilder()
            .setName(FIVETRAN_ACTIVE)
            .setType(DataType.BOOLEAN)
            .build();

        builder.addColumns(startCol);
        builder.addColumns(endCol);
        builder.addColumns(activeCol);

        return builder.build();
    }

    /**
     * Adds history mode columns to a table in the database.
     */
    public static void addHistoryModeColumnsToDb(DuckDBHelper dbHelper, String schema, String table) {
        List<Column> columnsToAdd = Arrays.asList(
            Column.newBuilder().setName(FIVETRAN_START).setType(DataType.UTC_DATETIME).build(),
            Column.newBuilder().setName(FIVETRAN_END).setType(DataType.UTC_DATETIME).build(),
            Column.newBuilder().setName(FIVETRAN_ACTIVE).setType(DataType.BOOLEAN).build()
        );

        for (Column column : columnsToAdd) {
            try {
                dbHelper.addColumn(schema, table, column);
            } catch (SQLException e) {
                logMessage(WARNING, "Failed to add column " + column.getName() + " to " + schema + "." + table + ": " + e.getMessage());
            }
        }
    }

    /**
     * Removes history mode columns from a table in the database.
     */
    public static void removeHistoryModeColumnsFromDb(DuckDBHelper dbHelper, String schema, String table) {
        List<String> columnsToDrop = Arrays.asList(FIVETRAN_START, FIVETRAN_END, FIVETRAN_ACTIVE);

        for (String column : columnsToDrop) {
            try {
                dbHelper.dropColumn(schema, table, column);
            } catch (SQLException e) {
                logMessage(WARNING, "Failed to drop column " + column + " from " + schema + "." + table + ": " + e.getMessage());
            }
        }
    }

    /**
     * Removes history mode columns from a table.
     */
    public static Table removeHistoryModeColumns(Table tableObj) {
        Table.Builder builder = Table.newBuilder(tableObj);
        builder.clearColumns();

        Set<String> historyColumns = new HashSet<>(Arrays.asList(FIVETRAN_START, FIVETRAN_END, FIVETRAN_ACTIVE));

        for (Column col : tableObj.getColumnsList()) {
            if (!historyColumns.contains(col.getName())) {
                builder.addColumns(col);
            }
        }

        return builder.build();
    }

    /**
     * Adds a soft delete column to a table.
     */
    public static Table addSoftDeleteColumn(Table tableObj, String columnName) {
        if (columnName == null || columnName.isEmpty()) {
            return tableObj;
        }

        Column softDelCol = Column.newBuilder()
            .setName(columnName)
            .setType(DataType.BOOLEAN)
            .build();

        return Table.newBuilder(tableObj)
            .addColumns(softDelCol)
            .build();
    }

    /**
     * Adds a soft delete column to a table in the database.
     */
    public static void addSoftDeleteColumnToDb(DuckDBHelper dbHelper, String schema, String table, String columnName) {
        if (columnName != null && !columnName.isEmpty()) {
            Column softDelCol = Column.newBuilder()
                .setName(columnName)
                .setType(DataType.BOOLEAN)
                .build();
            try {
                dbHelper.addColumn(schema, table, softDelCol);
            } catch (SQLException e) {
                logMessage(WARNING, "Failed to add soft delete column " + columnName + " to " + schema + "." + table + ": " + e.getMessage());
            }
        }
    }

    /**
     * Log a structured message.
     */
    private static void logMessage(String level, String message) {
        Gson gson = new Gson();
        Map<String, String> logEntry = new HashMap<>();
        logEntry.put("level", level);
        logEntry.put("message", message);
        logEntry.put("message-origin", "sdk_destination");
        System.out.println(gson.toJson(logEntry));
    }
}
